### PR TITLE
The Ventura Core Foundation Merge

### PR DESCRIPTION
**Pull request by iCharlesHu**
This PR brings the FOSS portions of Core Foundation on par with with macOS Ventura, iOS 16, watchOS 9 and tvOS 16. This includes the changes shipped during the previous releases.

Most of the changes include bug fixes and performance improvements, but some are useful to highlight:
- Updated Unicode definitions of `CFString` and `NSString`-based methods, including a new method of generating and storing this information by producing headers rather than bundling binary blobs.
- Updated `CFURLComponents` to support Internationalized Domain Names (IDNs) via Punycode encoding.
- Updates to calendrical calculation correctness.
- Improved performance in several spots throughout the framework.

Some development notes:
- `CF_CROSS_PLATFORM_EXPORT` has been renamed to `CF_EXPORT_NONOBJC_ONLY` to make its use clearer.
- Development has continued using primarily `os_*_lock_*` functions rather than `CFLock` or `CFMutex`. To avoid the codebase diverging too far and make future merges easier, this patch uses `os_unfair_lock` and similar facilities where they were previously replaced by `CFLock`. These types shim to `CFLock` implementations outside of Darwin.

### CODE REVIEW

**Comment by iCharlesHu** [comment]
@swift-ci please test

**Comment by iCharlesHu** [comment]
cc @parkera

**Comment by iCharlesHu** [comment]
cc @compnerd - Similar to previous merges, can we kick off a Windows build to see if there's anything breaks?

**Comment by AtariDreams** [comment]
@iCharlesHu I thought this was the source compiled at Apple for macOS?

File: CoreFoundation/RunLoop.subproj/CFRunLoop.c [hunk 1]
```diff
@@ -451,290 +447,12 @@ CF_INLINE kern_return_t __CFPortSetRemove(__CFPort port, __CFPortSet portSet) {
 CF_INLINE void __CFPortSetFree(__CFPortSet portSet) {
     close(portSet);
 }
-#elif TARGET_OS_BSD
```
Comment (3405691582) [inline CoreFoundation/RunLoop.subproj/CFRunLoop.c:454]: The changes from #3004 and #4622 need to be merged back in here.

**Review by 3405691582 — CHANGES_REQUESTED** [review]
I was going to add more finer grained review comments but I just did the needful anyway in #4648 -- feel free to pick those commits on this pr, since I don't know off the top of my head if I can chain prs. The comments about malloc introspection need to be addressed separately though, because it's not immediately apparent how to resolve that issue.

**Pull request closed without merging**